src/mender-update/standalone/states.cpp

```diff
@@ -0,0 +1,507 @@
+static const error::Error kFallbackError = error::MakeError(
+	error::ProgrammingError, "Returned from standalone operation without setting error code.");
+
+static void UpdateResult(ResultAndError &result, const ResultAndError &update) {
+	if (result.err == kFallbackError or result.err == error::NoError) {
+		result.err = update.err;
+	} else {
+		result.err = result.err.FollowedBy(update.err);
+	}
+
+	result.result = result.result | update.result;
+}
+
+void StateDataSaveState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	ctx.state_data.in_state = state_;
+
+	if (ResultContains(ctx.result_and_error.result, Result::Failed)) {
+		ctx.state_data.failed = true;
+	}
+	if (ResultContains(ctx.result_and_error.result, Result::RolledBack)
+		or ResultContains(ctx.result_and_error.result, Result::NoRollbackNecessary)) {
+		ctx.state_data.rolled_back = true;
+	}
+	if (ResultContains(ctx.result_and_error.result, Result::RollbackFailed)) {
+		ctx.state_data.rolled_back = false;
+	}
+
+	auto err = SaveStateData(ctx.main_context.GetMenderStoreDB(), ctx.state_data);
+	if (err != error::NoError) {
+		UpdateResult(ctx.result_and_error, {Result::Failed, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	poster.PostEvent(StateEvent::Success);
+}
+
+error::Error DoEmptyPayloadArtifact(Context &ctx) {
+	if (ctx.options != InstallOptions::NoStdout) {
+		cout << "Installing artifact..." << endl;
+		cout << "Artifact with empty payload. Committing immediately." << endl;
+	}
+
+	auto &data = ctx.state_data;
+	return ctx.main_context.CommitArtifactData(
+		data.artifact_name,
+		data.artifact_group,
+		data.artifact_provides,
+		data.artifact_clears_provides,
+		[](database::Transaction &txn) { return error::NoError; });
+}
+
+static io::ExpectedReaderPtr ReaderFromUrl(
+	events::EventLoop &loop, http::Client &http_client, const string &src) {
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	auto err = req->SetAddress(src);
+	if (err != error::NoError) {
+		return expected::unexpected(err);
+	}
+	error::Error inner_err;
+	io::AsyncReaderPtr reader;
+	err = http_client.AsyncCall(
+		req,
+		[&loop, &inner_err, &reader](http::ExpectedIncomingResponsePtr exp_resp) {
+			// No matter what happens, we will want to stop the loop after the headers
+			// are received.
+			loop.Stop();
+
+			if (!exp_resp) {
+				inner_err = exp_resp.error();
+				return;
+			}
+
+			auto resp = exp_resp.value();
+
+			if (resp->GetStatusCode() != http::StatusOK) {
+				inner_err = context::MakeError(
+					context::UnexpectedHttpResponse,
+					to_string(resp->GetStatusCode()) + ": " + resp->GetStatusMessage());
+				return;
+			}
+
+			auto exp_reader = resp->MakeBodyAsyncReader();
+			if (!exp_reader) {
+				inner_err = exp_reader.error();
+				return;
+			}
+			reader = exp_reader.value();
+		},
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
```

Comment:
It will, actually, but long after we have returned. The reader is a shared pointer and is returned from the function, so the body handler will be run once it resumes the loop and reads through the whole body. So it's important not to lambda-capture anything in this handler, since we are not in the same stack frame anymore.

I'll add a comment, since it's indeed quite unintuitive.